cmd/collectors/zapi/plugins/sensor/sensor.go

```diff
@@ -0,0 +1,295 @@
+				sensorUnit := instance.GetLabel("unit")
+				isAmbientMatch := ambientRegex.MatchString(sensorName)
+				isPowerMatch := powerRegex.MatchString(sensorName)
+				isVoltageMatch := voltageRegex.MatchString(sensorName)
+				isCurrentMatch := currentRegex.MatchString(sensorName)
+
+				if sensorType == "thermal" && isAmbientMatch {
+					if value, ok := metric.GetValueFloat64(instance); ok {
+						sensorEnvironmentMetricMap[iKey].ambientTemperature = append(sensorEnvironmentMetricMap[iKey].ambientTemperature, value)
+					}
+				}
+
+				if sensorType == "thermal" && !isAmbientMatch {
+					if value, ok := metric.GetValueFloat64(instance); ok {
+						sensorEnvironmentMetricMap[iKey].nonAmbientTemperature = append(sensorEnvironmentMetricMap[iKey].nonAmbientTemperature, value)
+					}
+				}
+
+				if sensorType == "fan" {
+					if value, ok := metric.GetValueFloat64(instance); ok {
+						sensorEnvironmentMetricMap[iKey].fanSpeed = append(sensorEnvironmentMetricMap[iKey].fanSpeed, value)
+					}
+				}
+
+				if isPowerMatch {
+					if value, ok := metric.GetValueFloat64(instance); ok {
+						if sensorEnvironmentMetricMap[iKey].powerSensor == nil {
+							sensorEnvironmentMetricMap[iKey].powerSensor = make(map[string]*sensorValue, 0)
+						}
+						sensorEnvironmentMetricMap[iKey].powerSensor[iKey2] = &sensorValue{name: iKey2, value: value, unit: sensorUnit}
+					}
+				}
+
+				if isVoltageMatch {
+					if value, ok := metric.GetValueFloat64(instance); ok {
+						if sensorEnvironmentMetricMap[iKey].voltageSensor == nil {
+							sensorEnvironmentMetricMap[iKey].voltageSensor = make(map[string]*sensorValue, 0)
+						}
+						sensorEnvironmentMetricMap[iKey].voltageSensor[iKey2] = &sensorValue{name: iKey2, value: value, unit: sensorUnit}
+					}
+				}
+
+				if isCurrentMatch {
+					if value, ok := metric.GetValueFloat64(instance); ok {
+						if sensorEnvironmentMetricMap[iKey].currentSensor == nil {
+							sensorEnvironmentMetricMap[iKey].currentSensor = make(map[string]*sensorValue, 0)
+						}
+						sensorEnvironmentMetricMap[iKey].currentSensor[iKey2] = &sensorValue{name: iKey2, value: value, unit: sensorUnit}
+					}
+				}
+
+			}
+		}
+	}
+
+	for _, k := range eMetrics {
+		my.createEnvironmentMetric(k)
+	}
+	for key, v := range sensorEnvironmentMetricMap {
+		instance, err := my.data.NewInstance(key)
+		instance.SetLabel("node", key)
+		for _, k := range eMetrics {
+			m := my.data.GetMetric(k)
+			switch k {
+			case "power":
+				var sumPower float64
+				if len(v.powerSensor) > 0 {
+					for _, v1 := range v.powerSensor {
+						if v1.unit == "mW" {
+							sumPower += sumPower + (v1.value / 1000)
+						} else if v1.unit == "W" {
+							sumPower += sumPower + v1.value
+						} else {
+							my.Logger.Warn().Str("unit", v1.unit).Str("value", fmt.Sprintf("%f", v1.value)).Msg("unknown power unit")
+						}
+					}
+				} else if len(v.voltageSensor) > 0 && len(v.voltageSensor) == len(v.currentSensor) {
```

Comment:
could be `else if len(v.voltageSensor) == len(v.currentSensor)` and we probably need an else block that warns when they are not equal

Reply:
len(v.voltageSensor) > 0 is to lot let this code block execute if voltage , current sensor size is zero

Reply:
we don't need to if size is zero. I have added size in warning log.